### PR TITLE
[RW-5681][risk=no] Refactor Puppeteer job in CircleCI config file

### PR DESCRIPTION
**Pull request by aweng98**
If "SKIP E2E" string (case-insensitive) is present in a PR commit message, CircleCI will not run `puppeteer-e2e-test` job.

Halted job screenshot:
Screen Shot 2020-09-30 at 3 12 07 PM

Continued job screenshot:
Screen Shot 2020-09-30 at 3 14 36 PM

Also, single `puppeteer-e2e-test` job to run for PR commits and master branch merge.

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -601,31 +625,19 @@ jobs:
     parallelism: << parameters.parallel_num >>
     steps:
       - checkout_init_git
-      - browser-tools/install-browser-tools
+      - when:
+          # Deploy local UI server then run Puppeteer tests against "local ui" + "test api" environment.
+          condition: << parameters.require_local_ui >>
+          steps:
+            - skip_e2e_commit
```
Comment (calbach) [inline .circleci/config.yml:632]: A few issues with this approach:

- Skipping on non-PR, skipping on e2e commit and skipping on lack of code change: these are all presubmit optimizations. While it's true that you also only want to run against a local UI on presubmit, it's not directly  related to local UI testing.
- There's now two parameters which overlap and could be provided inconsistently: `env_name` and `require_local_ui`

If you want to keep one job, I'd suggest renaming this param to `presubmit` or `is_presubmit`. For `start_local_ui`, I would put this in a separate conditional which is dependent only on `<<env_name>> == "local"`. Alternatively, you could avoid these issues by reverting back to two jobs, though this does result in some duplication

File: .circleci/config.yml [hunk 2]
```diff
@@ -291,6 +298,20 @@ commands:
           name: Wait for local UI server to start
           command: dockerize -wait tcp://localhost:4200 -timeout 4m
 
+  skip_e2e_commit:
+    description: "Halt job if commit message contains 'SKIP E2E'"
```
Comment (calbach) [inline .circleci/config.yml:302]: would specify somewhere here that this is case insensitive. The bash `,,` is nice but also pretty obscure, I had to look this up.

File: .circleci/config.yml [hunk 3]
```diff
@@ -291,6 +298,20 @@ commands:
...
+    steps:
+      - run:
+          name: "Halt job if commit message contains 'SKIP E2E'"
```
Comment (calbach) [inline .circleci/config.yml:305]: On the subject of the "skip e2e" behavior - I agree with this approach, that the tests should be run by default.

File: .circleci/config.yml (hunk 2 above)
Comment (aweng98) [reply .circleci/config.yml:302]: Added a comment to explain what the double comma is called.

File: .circleci/config.yml (hunk 3 above)
Comment (aweng98) [reply .circleci/config.yml:305]: Cool.

**Review by aweng98 — COMMENTED** [review]
Hoping to avoid two jobs with duplicated steps, so I changed code:
-  Created a new command called `e2e_presubmit_check` which includes all presubmit optimizations.
-  So I don't know how to add a conditional which is dependent on `<< parameters.env_name >> == "local"`. But I
learned job parameter `type: steps`. When `env_name ==  local`, this new parameter includes `e2e_presubmit_check` and `start_local_ui` jobs. Hopefully this approach makes obvious if parameters are used inconsistently.

File: .circleci/config.yml [hunk 4]
```diff
@@ -213,6 +219,7 @@ commands:
         enum: ["local", "test", "staging"]
     steps:
       - install_e2e_dependencies
+      - browser-tools/install-browser-tools
```
Comment (jaycarlton) [inline .circleci/config.yml:222]: Some steps use hyphens and others use underscores. I'd pick one style and stick with it. Underscores seem  safer to me, and more standard as identifiers in  mosts contexts.

File: .circleci/config.yml (hunk 2 above)
Comment (jaycarlton) [reply .circleci/config.yml:302]: Thinking out loud: could we have a generic filter like `skip <task_name>`?

File: .circleci/config.yml (hunk 3 above)
Comment (jaycarlton) [reply .circleci/config.yml:305]: aside:  Eventually it might be nice to encapsulate all the trigger evaluation and job management into  a single "driver" task, and have separate configs for the real  work.

It might also be  worth looking at a small e2e smoke test suite that runs quickly and still gives reasonable integration coverage. Unless managing the environment is a  huge contributor to the time.

File: .circleci/config.yml [hunk 5]
```diff
@@ -291,6 +298,30 @@ commands:
...
+    description: "Halt job if commit message contains 'SKIP E2E'"
+    steps:
+      - run:
+          name: "Halt job if commit message contains 'SKIP E2E'"
+          command: |
+            echo 'export COMMIT_MESSAGE=$(git log -1 --pretty=format:"%s")' >> $BASH_ENV
+            source $BASH_ENV
+            echo "Git commit message: $COMMIT_MESSAGE"
+            # Double comma is called "Parameter Expansion" in Bash version 4+. It converts string to lowercase letters.
+            if [[ "${COMMIT_MESSAGE,,}" == *"skip e2e"* ]]; then
```
Comment (jaycarlton) [inline .circleci/config.yml:311]: FWIW, it might be cleaner to have  a little Ruby or Python  script to parse the commit message &  other metadata  and give a cleaner structure. (You could even  write test cases for it). My tolerance for reading shell scripts is about 2-1/2 lines.

File: .circleci/config.yml [hunk 6]
```diff
@@ -291,6 +298,30 @@ commands:
...
+            fi
+
+
+  e2e_presubmit_check:
+    steps:
+      - e2e_skip_commit
+      - halt_if_code_unchanged
+      - run:
+          name: "Halt job if not a pull request"
+          command: bash .circleci/pr-skip-ci.sh
```
Comment (jaycarlton) [inline .circleci/config.yml:323]: Do  all the halt commnads run here?

Also, can we  mark the test as not run instead of passed or failed somehow? (scope creep, I know). There's a nice gray X or something  in the UI, but I don't  remember what triggered it.

File: .circleci/config.yml [hunk 7]
```diff
@@ -77,6 +77,12 @@ anchors:
       # Workbench release tag format example: v5-3-rc1
       only: /^v[0-9]+-[0-9]+-rc[0-9]+$/
 
+  # Job runs for PR branches only
+  filter_only_pr_branch: &filter_only_pr_branch
+    tags:
+      ignore: /.*/
```
Comment (jaycarlton) [inline .circleci/config.yml:83]: I don't  understand what this does. Is it ignoring hidden files (like a glob), or is this a regex (so  it's  ignoring everything)?

Comment (aweng98) [reply .circleci/config.yml:83]: Igore all tag branches (include release tag branches).

File: .circleci/config.yml (hunk 4 above)
Comment (aweng98) [reply .circleci/config.yml:222]: Good observation. circleci seems to use hyphens for everything. We use underscores. If we want to keep all consistent, we have to change to hyphen to match. It would be in a separate PR.

File: .circleci/config.yml (hunk 2 above)
Comment (aweng98) [reply .circleci/config.yml:302]: Possible. One way I know how to do that is make a common command to store the commit message then each job will parse and pattern match the message.
e.g.
```

  get_commit_message:
    steps:
      - run:
          command: |
            echo 'export COMMIT_MESSAGE=$(git log -1 --pretty=format:"%s")' >> $BASH_ENV
            source $BASH_ENV
          
  e2e_presubmit_check:
    steps:
      - get_commit_message
      - run:
          - command: |
              if [[ "${COMMIT_MESSAGE,,}" == *"skip e2e"* ]]; then
                echo "Halting e2e test."
                circleci step halt
              fi
```

File: .circleci/config.yml [hunk 8]
```diff
@@ -647,9 +659,18 @@ workflows:
           filters: *filter_only_master_branch
           requires:
             - ui-unit-test
+      # Deploy local ui server then run Puppeteer tests against "local ui" + "test api" environment for PR commits only.
+      - puppeteer-e2e-test:
+          filters: *filter_only_pr_branch
+          env_name: "local"
+          parallel_num: 2
+          optional_steps:
```
Comment (calbach) [inline .circleci/config.yml:667]: This is interesting - I like this for the presubmit check. But still it would be ideal if `env_name` and `start_local_ui` were not provided separately.

You should be able to make the start_local_ui step run conditionally as I was suggesting earlier. You can do a comparison as described here: https://circleci.com/docs/2.0/configuration-reference/#logic-statements

```
when:
  condition:
    equal: [ local, << parameters.env_name >>]
  steps:
    - run_local_ui
```

Comment (aweng98) [reply .circleci/config.yml:678]: WOW. It works. I didn't know that. Thanks.

File: .circleci/config.yml (hunk 4 above)
Comment (aweng98) [reply .circleci/config.yml:222]: @jaycarlton ticket to track hyphen refactor. [RW-5689](https://precisionmedicineinitiative.atlassian.net/browse/RW-5689)

**Review by calbach — APPROVED** [review]
Nice
